### DISCUSSION TITLE
Sending executable files via python UART

### DISCUSSION THREAD

**Discussion by Giorgio122** [stnolting/neorv32#1075] [Q&A] [answered]
Goodmorning everyone,
I am currently working with the neorv and what i need to do is to automatize a procedure in which i need to send the executable via UART through the bootloader. For this purpose I have written a python script interacting with the neorv UART. Everything is working except when i try to send the .bin file.

def send_bin_file(file_to_send):
	with open(file_to_send,'rb') as file:
		while True:
			data = file.read()
			ser.write(data)
			time.sleep(0.1)

This is the code I am using to send the data. I have also tried to modify this into different ways but none of them seems to be working. So, first of all, is it impossible to send the data or did someone already tried and find a way to transmit the executable via python UART?

Thanks!

**Comment by stnolting** [comment]
Hey @Giorgio122!

I'm sorry, but I have no idea how to do that with Python... 🙈

This is what GH AI says about it:

```python
import serial

# Open the serial port
ser = serial.Serial('/dev/ttyUSB0', 19200)  # Replace '/dev/ttyUSB0' with your serial port

# Read the binary file
with open('path/to/your/binary/file.bin', 'rb') as file:
    binary_data = file.read()

# Send the binary data
ser.write(binary_data)

# Close the serial port
ser.close()
```

Maybe someone else (a human) can help?

**Reply by Giorgio122** [reply]
Hello @stnolting,
I have already tried also something like that, unfortunately every time the neorv return ERR_EXE

**Reply by stnolting** [reply, marked as answer]
Before sending the actual executable you need to send two more characters: one arbitrary one to skip the auto-boot sequence and an `u` for starting the upload.

We still have a bash script for automatic executable upload: [`sw/image_gen/uart_upload.sh`](https://github.com/stnolting/neorv32/blob/main/sw/image_gen/uart_upload.sh)

Maybe you can just call that from your Python code?

**Reply by Giorgio122** [reply]
That is a possibility. I knew about the other chars but that is not an issue. Maybe launching the script could be the solution, just slightly modifying it and adapt it to my need. Thanks, I will let you know the outcome!
